vm/src/warn.rs

```diff
@@ -31,19 +78,172 @@ pub fn warn(
+        }?;
+
+        /* Python code: action, msg, cat, mod, ln = item */
+        let action = if let Some(action) = tmp_item.get(0) {
+            action.str(vm).map(|action| action.into_object())
+        } else {
+            Err(vm.new_type_error("action must be a string".to_string()))
+        };
+
+        let good_msg = if let Some(msg) = tmp_item.get(1) {
+            check_matched(msg, &text, vm)?
+        } else {
+            false
+        };
+
+        let is_subclass = if let Some(cat) = tmp_item.get(2) {
+            category.fast_isinstance(&cat.class())
+        } else {
+            false
+        };
+
+        let good_mod = if let Some(item_mod) = tmp_item.get(3) {
+            check_matched(item_mod, &module, vm)?
+        } else {
+            false
+        };
+
+        let ln = tmp_item.get(4).map_or_else(
+            || 0,
+            |ln_obj| {
+                if let Ok(ln) = ln_obj.try_int(vm) {
+                    ln.as_u32_mask() as usize
+                } else {
+                    0
+                }
+            },
+        );
+
+        if good_msg && good_mod && is_subclass && (ln == 0 || lineno == ln) {
+            _item = tmp_item;
+            return action;
+        }
+    }
+
+    get_default_action(vm)
+}
+
+fn already_warned(
+    registry: PyObjectRef,
+    key: PyObjectRef,
+    should_set: bool,
+    vm: &VirtualMachine,
+) -> PyResult<bool> {
+    let version_obj = registry.get_item(identifier!(&vm.ctx, version), vm).ok();
+    let filters_version = vm.ctx.new_int(vm.state.warnings.filters_version).into();
+
+    match version_obj {
+        Some(version_obj)
+            if version_obj.try_int(vm).is_ok() || version_obj.is(&filters_version) =>
+        {
+            let already_warned = registry.get_item(key.as_ref(), vm)?;
+            if already_warned.is_true(vm)? {
+                return Ok(true);
+            }
+        }
+        _ => {
+            let registry = registry.dict();
+            registry.as_ref().map(|registry| {
+                registry.clear();
+                registry
+            });
+
+            if let Some(registry) = registry {
+                if registry.set_item("version", filters_version, vm).is_err() {
+                    return Ok(false);
+                }
+            }
+        }
+    }
+
+    /* This warning wasn't found in the registry, set it. */
+    Ok(if should_set {
+        let item = vm.ctx.true_value.clone().into();
+        registry.set_item(key.as_ref(), item, vm).map(|_| true)?
+    } else {
+        false
+    })
+}
+
+fn normalize_module(filename: PyStrRef, vm: &VirtualMachine) -> Option<PyObjectRef> {
+    let len = filename.char_len();
+
+    if len == 0 {
+        Some(vm.new_pyobj("<unknown>"))
+    } else if len >= 3 && filename.as_str().ends_with(".py") {
+        Some(vm.new_pyobj(&filename.as_str()[..len - 3]))
+    } else {
+        Some(filename.as_object().to_owned())
+    }
```

Comment:
match works better for this kind of case

```suggestion
    let obj = match filename.char_len() {
        0 => vm.new_pyobj("<unknown>"),
        len if len >= 3 && filename.as_str().ends_with(".py") => {
            vm.new_pyobj(&filename.as_str()[..len - 3])
        }
        _ => filename.as_object().to_owned(),
    };
    Some(obj)
```